### PR TITLE
TST: Refactored Ad GridOperator tests.

### PR DESCRIPTION
**Pull request by keileg**
Group the tests for subdomain, mortar and boundary projections into one class each, and let them share setup through fixtures and parametrization instead of repeating it. The expected projection matrices are assembled by helper functions that are independent of the implementation under test.

This PR was extracted from #1747 and must be merged before review of that PR can be finalized.

### CODE REVIEW

File: tests/numerics/ad/test_grid_operators.py [hunk 1]
```diff
@@ -28,192 +32,243 @@ def mdg():
...
-
-    # Test projections to and from an empty list of subdomains.
-    assert proj.cell_restriction([]).shape == (0, n_cells)
-    assert proj.cell_prolongation([]).shape == (n_cells, 0)
-    assert proj.face_restriction([]).shape == (0, n_faces)
-    assert proj.face_prolongation([]).shape == (n_faces, 0)
-
-    # Test projection of one fracture at a time for the full set of grids.
-    for sd in subdomains:
+def _known_subdomain_projection_matrix(subdomains, target_grids, proj_dim, entity):
```
Comment (IvarStefansson) [inline tests/numerics/ad/test_grid_operators.py]: Please add at least one of typing and parameter descriptions. The docstring is really dense.

File: tests/numerics/ad/test_grid_operators.py [hunk 2]
```diff
@@ -28,192 +32,243 @@ def mdg():
...
-        (np.hstack((dc1, dc2)), (np.hstack((rc1, rc2)), np.hstack((cc1, cc2)))),
-        shape=(num_rows_cell, n_cells),
-    ).tocsr()
-    known_face_proj = sps.coo_matrix(
-        (np.hstack((df1, df2)), (np.hstack((rf1, rf2)), np.hstack((cf1, cf2)))),
-        shape=(num_rows_face, n_faces),
-    ).tocsr()
+    Covers three scenarios (one per test method below): projecting to/from an empty
+    grid list, a single grid, and a combined list of two grids. Each scenario is
+    checked both for numerical correctness of the projection matrix (except the
```
Comment (IvarStefansson) [inline tests/numerics/ad/test_grid_operators.py]: The "both" seems to miss an "and" clause.

File: tests/numerics/ad/test_grid_operators.py [hunk 3]
```diff
@@ -28,192 +32,243 @@ def mdg():
...
+        expected = known if is_restriction else known.T
+
+        op = getattr(proj, method_name)([g1, g2])
+
+        assert _compare_matrices(op, expected)
+
+
+# MARK: Mortar<->primary/secondary subdomain projections ---------------------------
+
+#: (method_name, pair, kind, is_to_mortar) for each of the eight mortar<->primary/
```
Comment (IvarStefansson) [inline tests/numerics/ad/test_grid_operators.py]: Why is there a colon after the hash?

File: tests/numerics/ad/test_grid_operators.py [hunk 4]
```diff
@@ -229,235 +284,331 @@ def _col_offset(interfaces):
...
+                g_new = pp.refinement.refine_grid_1d(g, ratio=2)
+                mdg.replace_subdomains_and_interfaces({g: g_new})
+
+        proj = pp.ad.MortarProjections(
+            subdomains=mdg.subdomains(), interfaces=mdg.interfaces(), mdg=mdg, dim=1
+        )
+        integrated = getattr(proj, int_method)()
+        # A second call returns the stored matrix rather than constructing a new one.
+        assert getattr(proj, int_method)() is integrated
+
```
Comment (IvarStefansson) [inline tests/numerics/ad/test_grid_operators.py]:
```suggestion
    # Separate "secondary" methods iff non_matching. Negate to get shared.
```

File: tests/numerics/ad/test_grid_operators.py [hunk 5]
```diff
@@ -229,235 +284,331 @@ def _col_offset(interfaces):
...
+        method_name,
+        pair,
+        kind,
+        is_to_mortar,
+    ):
+        """Mortar<->primary/secondary subdomain projections match matrices assembled
+        independently from each interface's own local mortar_to_primary/secondary_
+        int/avg matrices.
+
+        The X_to_mortar directions are the transpose of the mortar_to_X direction,
```
Comment (IvarStefansson) [inline tests/numerics/ad/test_grid_operators.py:500]: Is this true for non_matching?

File: tests/numerics/ad/test_grid_operators.py [hunk 6]
```diff
@@ -229,235 +284,331 @@ def _col_offset(interfaces):
...
+        # Compute geometry for the mixed-dimensional grid. This is needed for the
+        # boundary projection operator.
+        mdg.compute_geometry()
+        return mdg.subdomains()
+
+    @pytest.fixture
+    def projection(self, mdg, subdomains, proj_dim):
+        return pp.ad.BoundaryProjection(mdg, subdomains, proj_dim)
+
+    def test_shape(self, mdg, proj_dim, projection):
```
Comment (IvarStefansson) [inline tests/numerics/ad/test_grid_operators.py]: Is "test_*" a bad name for a non-test helper method?

File: tests/numerics/ad/test_grid_operators.py [hunk 7]
```diff
@@ -229,235 +284,331 @@ def _col_offset(interfaces):
...
+        sides), and 0 for the 0d intersection (trivially, since it has no faces).
+        """
+        subdomain_to_boundary = projection.subdomain_to_boundary.parse(mdg)
+        starts = np.cumsum(
+            np.hstack((0, [sd.num_faces * proj_dim for sd in subdomains]))
+        )
+        block = subdomain_to_boundary[:, starts[sd_index] : starts[sd_index + 1]]
+        assert np.sum(block) == expected_sum_factor * proj_dim
+
+    def test_subset_of_grids(self, mdg, proj_dim):
```
Comment (IvarStefansson) [inline tests/numerics/ad/test_grid_operators.py]: I think this test could have been split. Feel free to disagree or agree in principle, but not split this particular one - it's a minor detail.

**Review by IvarStefansson — APPROVED** [review]
Looks good in the main. Disclaimer: Didn't parse everything in detail.

File: tests/numerics/ad/test_grid_operators.py (hunk 5 above)
Comment (keileg) [reply tests/numerics/ad/test_grid_operators.py:500]: Yes. Please never ask how much time I have spent thinking of this.

File: tests/numerics/ad/test_grid_operators.py (hunk 6 above)
Comment (keileg) [reply tests/numerics/ad/test_grid_operators.py]: Deleted the whole method, it was not needed.
